src/sage/rings/polynomial/generalized_order.py

```diff
@@ -0,0 +1,428 @@
+            sage: G.group_order()
+            Lexicographic term order
+        """
+        return self._group_order
+
+    def group_order_name(self):
+        r"""
+        Return the name of the underlying group order.
+
+        EXAMPLES::
+
+            sage: from sage.rings.polynomial.generalized_order import GeneralizedOrder
+            sage: order = GeneralizedOrder(2,group_order="lex")
+            sage: order.group_order_name()
+            'lex'
+        """
+        return self._group_order_name
+    
+    def score_function_name(self):
+        r"""
+        Return the name of the underlying score function.
+
+        EXAMPLES::
+
+            sage: from sage.rings.polynomial.generalized_order import GeneralizedOrder
+            sage: order = GeneralizedOrder(2,score_function="degmin")
+            sage: order.score_function_name()
+            'degmin'
+        """
+        return self._score_function_name
+
+    def score(self,t):
+        r"""
+        Compute the score of a tuple ``t``.
+
+        INPUT:
+
+        - `t` -- a tuple of integers
+
+        EXAMPLES::
+
+            sage: from sage.rings.polynomial.generalized_order import GeneralizedOrder
+            sage: order = GeneralizedOrder(2)
+            sage: order.score((-2,3))
+            2
+        """
+        return self._score_function(t)
+                                
+    def compare(self,a,b):
+        r"""
+        Return 1, 0 or -1 whether tuple ``a`` is greater than, equal or less than to tuple ``b`` respectively.
+
+        INPUT:
+
+        - `a` and `b` -- two tuples of integers
+
+        EXAMPLES::
+
+            sage: from sage.rings.polynomial.generalized_order import GeneralizedOrder
+            sage: order = GeneralizedOrder(3)
+            sage: order.compare((1,2,-2), (3,-4,5))
+            -1
+            sage: order.compare((3,-4,5), (1,2,-2))
+            1
+            sage: order.compare((3,-4,5), (3,-4,5))
+            0
+        """
+        if a == b: return 0
+        diff = self._score_function(a) - self._score_function(b)
+        if diff != 0: 
+            return 1 if diff > 0 else -1
+        else: 
+            return 1 if self._group_order.greater_tuple(a,b) == a else -1
+
+    def greatest_tuple(self,*L):
+        r"""
+        Return the greatest tuple in ``L``.
+
+        INPUT:
+
+        - *L -- integer tuples
+
+        EXAMPLES::
+
+            sage: from sage.rings.polynomial.generalized_order import GeneralizedOrder
+            sage: order = GeneralizedOrder(3)
+            sage: order.greatest_tuple((0,0,1),(2,3,-2))
+            (2, 3, -2)
+            sage: L = [(1,2,-1),(3,-3,0),(4,-5,-6)]
+            sage: order.greatest_tuple(*L)
+            (4, -5, -6)
+        """
+        n = len(L) 
+        if n == 0:
+           raise ValueError("empty list of tuples")
+
+        if n == 1:
+            return L[0]
+        else:
+            a = L[0]
+            b = self.greatest_tuple(*L[1:])
```

Comment:
I don't really like this recursive call because it implies all useless checkings.
Why not programming this imperatively with a for loop?